tests/31sync/06state.pl

```diff
@@ -403,3 +461,69 @@
          Future->done(1);
       })
    };
+
+
+test "When user joins and leaves a room in the same batch, the full state is still included in the next sync",
+   bug => 'SYN-514',
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user_a, $filter_id_a, $user_b, $filter_id_b, $room_id, $next_b );
+
+      my $filter = {
+         room => {
+            timeline  => { types => [] },
+            state     => { types => [ "a.madeup.test.state" ] },
+            ephemeral => { types => [] },
+         },
+         presence => { types => [] },
+      };
+
+      Future->needs_all(
+         matrix_register_user_with_filter( $http, $filter ),
+         matrix_register_user_with_filter( $http, $filter ),
+      )->then( sub {
+         ( $user_a, $filter_id_a, $user_b, $filter_id_b ) = @_;
+
+         matrix_create_room( $user_a );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_put_room_state( $user_a, $room_id,
+            type      => "a.madeup.test.state",
+            content   => { "my_key" => 1 },
+            state_key => "",
+         );
+      })->then( sub {
+         matrix_invite_user_to_room( $user_a, $user_b, $room_id );
+      })->then( sub {
+         matrix_sync( $user_b, filter => $filter_id_b );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         $next_b = $body->{next_batch};
+         matrix_join_room( $user_b, $room_id );
+      })->then( sub {
+         matrix_leave_room( $user_b, $room_id );
+      })->then( sub {
+         matrix_sync( $user_b, filter => $filter_id_b, since => $next_b );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id};
+         require_json_keys( $room, qw( event_map timeline state ephemeral ));
+         @{ $room->{state}{events} } == 1
```

Comment:
This might be a bit neater as:

    my $eventcount = scalar @{ $room->{state}{events} };
    $eventcount == 1 or
       die "Expected one state event, got $eventcount";